syntax = "proto3";

package raven;

import "market_data.proto";

service MarketDataService {
    rpc StreamMarketData(stream SubscriptionRequest) returns (stream MarketDataMessage);
    rpc Subscribe(SubscribeRequest) returns (SubscribeResponse);
    rpc Unsubscribe(UnsubscribeRequest) returns (UnsubscribeResponse);
    rpc GetHistoricalData(HistoricalDataRequest) returns (stream MarketDataMessage);
}

message SubscriptionRequest {
    oneof request {
        SubscribeRequest subscribe = 1;
        UnsubscribeRequest unsubscribe = 2;
        HeartbeatRequest heartbeat = 3;
    }
}

message SubscribeRequest {
    string client_id = 1;
    repeated string symbols = 2;
    repeated DataType data_types = 3;
    map<string, string> filters = 4;
}

message UnsubscribeRequest {
    string client_id = 1;
    repeated string symbols = 2;
    repeated DataType data_types = 3;
}

message SubscribeResponse {
    bool success = 1;
    string message = 2;
    repeated string subscribed_topics = 3;
}

message UnsubscribeResponse {
    bool success = 1;
    string message = 2;
    repeated string unsubscribed_topics = 3;
}

message HeartbeatRequest {
    string client_id = 1;
    int64 timestamp = 2;
}

message HistoricalDataRequest {
    repeated string symbols = 1;
    repeated DataType data_types = 2;
    int64 start_time = 3;
    int64 end_time = 4;
    int32 limit = 5;
}

enum DataType {
    ORDERBOOK = 0;
    TRADES = 1;
    CANDLES_1M = 2;
    CANDLES_5M = 3;
    CANDLES_1H = 4;
    FUNDING_RATES = 5;
    WALLET_UPDATES = 6;
}